File: js/data-registry.js
/** Client-side catalog loader for racing titles and JSON datasets. */
(function (global) {
  const DEFAULT_GAME_ID = "ac_evo";

  class DataRegistry {
    constructor(manifest, basePath = "data") {
      this.manifest = manifest;
      this.basePath = basePath.replace(/\/$/, "");
      this._cache = {};
    }

    listGames() {
      return (this.manifest.games || []).filter((g) => g.active !== false);
    }

    getGame(gameId) {
      return (this.manifest.games || []).find((g) => g.id === gameId);
    }

    async loadDataset(gameId, dataset) {
      const key = `${gameId}:${dataset}`;
      if (this._cache[key]) return this._cache[key];
      const game = this.getGame(gameId);
      if (!game) throw new Error(`Unknown game: ${gameId}`);
      const filename = game.data_files?.[dataset];
      if (!filename) throw new Error(`Game ${gameId} has no ${dataset} dataset`);
      const res = await fetch(`${this.basePath}/${filename}`);
      if (!res.ok) throw new Error(`Failed to load ${filename}`);
      const data = await res.json();
      this._cache[key] = data;
      return data;
    }

    async getCars(gameId = DEFAULT_GAME_ID) {
      const cars = await this.loadDataset(gameId, "cars");
      return [...cars].sort((a, b) => a.name.localeCompare(b.name));
    }

    async getTracks(gameId = DEFAULT_GAME_ID) {
      const tracks = await this.loadDataset(gameId, "tracks");
      return [...tracks].sort((a, b) => {
        const c = a.name.localeCompare(b.name);
        return c !== 0 ? c : (a.layout || "").localeCompare(b.layout || "");
      });
    }

    async findCar(carId, gameId = DEFAULT_GAME_ID) {
      const cars = await this.getCars(gameId);
      return cars.find((c) => c.id === carId) || null;
    }

    async findTrack(trackId, gameId = DEFAULT_GAME_ID) {
      const tracks = await this.getTracks(gameId);
      return tracks.find((t) => t.id === trackId) || null;
    }
  }

  async function createDataRegistry(basePath = "data") {
    const res = await fetch(`${basePath}/manifest.json`);
    if (!res.ok) throw new Error("Failed to load game manifest");
    const manifest = await res.json();
    return new DataRegistry(manifest, basePath);
  }

  global.DEFAULT_GAME_ID = DEFAULT_GAME_ID;
  global.DataRegistry = DataRegistry;
  global.createDataRegistry = createDataRegistry;
})(window);

File: data_registry.py
"""Modular catalog loader for racing titles and their car/track datasets."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

DATA_DIR = Path(__file__).parent / "data"
DEFAULT_GAME_ID = "ac_evo"


class DataRegistry:
    """Loads game manifests and per-title JSON catalogs with in-memory caching."""

    def __init__(self, data_dir: Path | None = None) -> None:
        self.data_dir = data_dir or DATA_DIR
        self._manifest: dict[str, Any] | None = None

    @property
    def manifest(self) -> dict[str, Any]:
        if self._manifest is None:
            with open(self.data_dir / "manifest.json", encoding="utf-8") as f:
                self._manifest = json.load(f)
        return self._manifest

    def list_games(self) -> list[dict[str, Any]]:
        return [g for g in self.manifest.get("games", []) if g.get("active", True)]

    def get_game(self, game_id: str) -> dict[str, Any] | None:
        return next((g for g in self.manifest.get("games", []) if g["id"] == game_id), None)

    def _load_dataset(self, game_id: str, dataset: str) -> list[dict[str, Any]]:
        game = self.get_game(game_id)
        if not game:
            raise ValueError(f"Unknown game: {game_id}")
        filename = game.get("data_files", {}).get(dataset)
        if not filename:
            raise ValueError(f"Game {game_id} has no {dataset} dataset")
        path = self.data_dir / filename
        with open(path, encoding="utf-8") as f:
            return json.load(f)

    def get_cars(self, game_id: str = DEFAULT_GAME_ID) -> list[dict[str, Any]]:
        return sorted(self._load_dataset(game_id, "cars"), key=lambda c: c["name"])

    def get_tracks(self, game_id: str = DEFAULT_GAME_ID) -> list[dict[str, Any]]:
        return sorted(
            self._load_dataset(game_id, "tracks"),
            key=lambda t: (t["name"], t.get("layout", "")),
        )

    def find_car(self, car_id: str, game_id: str = DEFAULT_GAME_ID) -> dict[str, Any] | None:
        return next((c for c in self.get_cars(game_id) if c["id"] == car_id), None)

    def find_track(self, track_id: str, game_id: str = DEFAULT_GAME_ID) -> dict[str, Any] | None:
        return next((t for t in self.get_tracks(game_id) if t["id"] == track_id), None)


@lru_cache(maxsize=1)
def default_registry() -> DataRegistry:
    return DataRegistry()
